sdk/src/main/java/io/opentelemetry/sdk/metrics/BoundRegistry.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.metrics;
+
+import io.opentelemetry.metrics.LabelSet;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+
+abstract class BoundRegistry<B extends AbstractBoundInstrument> {
+  private final ConcurrentHashMap<LabelSet, B> bounds;
+  private final ReentrantLock collectLock;
+
+  BoundRegistry() {
+    bounds = new ConcurrentHashMap<>();
+    collectLock = new ReentrantLock();
+  }
+
+  B acquire(LabelSet labelSet) {
+    B bound = bounds.get(labelSet);
+    if (bound != null && bound.ref()) {
+      // At this moment it is guaranteed that the Bound is in the map and will not be removed.
+      return bound;
+    }
+
+    // Missing entry or no longer mapped, try to add a new entry.
+    bound = newBound(labelSet);
+    while (true) {
+      B oldBound = bounds.putIfAbsent(labelSet, bound);
+      if (oldBound != null) {
+        if (oldBound.ref()) {
+          // At this moment it is guaranteed that the Bound is in the map and will not be removed.
+          return oldBound;
+        }
+        // Try to remove the oldBound. This will race with the collect method, but only one will
+        // succeed.
+        bounds.remove(labelSet, oldBound);
+        continue;
+      }
+      return bound;
```

Comment:
I am not a huge fan of while(true) loops. I would use a defensive programming design, capping the max exec time of the while. Something like: 
```java 
do {
...
} while(timer > 0)
```
I am not sold on this, though. What is your opinion?

Reply:
I am not worried, as you can see we try to remove the value on the critical path if an unmapped value is still in the map, so the while loop is not going to spin more than twice.

Reply:
I'd definitely like to avoid putting in a timer as well. I'm ok with the spin loop. Is there a way to write tests the various scenarios here? Maybe a small stress test would help to alleviate concerns.